### PR TITLE
Writing tutorial

### PR DESCRIPTION
**Pull request by Huy-Ngo**

### CODE REVIEW

File: src/tutorial/context.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
+Context creation
+==================
+
+A context is an object that allows palace to access OpenAL,
+which is essential when you work with palace. Context maintains
+the audio environment and contains environment settings and components
+such as sources, buffers, and effects.
+
+Creating a device object
+---------------------------
```
Comment (McSinyx) [inline src/tutorial/context.rst]: Please make sure the under line is under something.

File: src/tutorial/context.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+To create a context, we must first create a device,
+since it's a parameter of the context object.
+
+To create an object, well, you just have to instantiate
+the `Device` class.
+
+.. highlight:: python
```
Comment (McSinyx) [inline src/tutorial/context.rst]: Please use `code-block` instead.

File: src/tutorial/context.rst [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+This is how you declare a `Device` object with the default device.
+There can be several devices available, which can be found
+in `palace.device_names`.
+
+Creating a context
+--------------------
+
+Now that we've created a device, we can create the context:
+
+
```
Comment (McSinyx) [inline src/tutorial/context.rst:37]: Pay attention to spacing.

File: src/tutorial/context.rst [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+           # Your code goes here
+
+Before we continue, if we only use `dev` for context creation,
+is there a shorter way to write it? The answer is yes.
+
+.. highlight:: python
+   from palace import Device, Context
+
+   with Device(), Context() as ctx:
+           # Your code goes here
```
Comment (McSinyx) [inline src/tutorial/context.rst]: Note the indentation here.

**Review by McSinyx — COMMENTED** [review]
Some early nits.

File: src/tutorial/context.rst (hunk 1 above)
Comment (Huy-Ngo) [reply src/tutorial/context.rst]: Hmm... the syntax highlighting enlarged the text so I couldn't see it.

File: src/tutorial/context.rst (hunk 3 above)
Comment (McSinyx) [reply src/tutorial/context.rst:37]: There should be a newline after this, however sphinx would tell you when you build so...

**Comment by McSinyx** [comment]
IMHO you should keep the documentation process incremental.  While there's not much to review, it'd still be difficult to review more than a few pages of tutorial at a time.

**Comment by Huy-Ngo** [comment]
By incremental you mean adding only one or a few pages at a time?

Since I already started on writing the second tutorial, is there a way to merge only a few commits?

**Comment by McSinyx** [comment]
Try not adding new pages and finish the existing ones.

File: src/tutorial/play-audio.rst [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+.. code-block:: python
+
+   from palace import Device, Context, Source, decode
+
+   fname = 'some_audio.ogg'
+
+   with Device(), Context() as ctx:
+       with Source() as src:
+           dec = decode(fname)
+	   dec.play(12000, 4, src)
```
Comment (McSinyx) [inline src/tutorial/play-audio.rst]: Please pay attention to indentation.  Too bad I'm not aware of any linter that can do this.  Also I feel the newline between the `fname` (use `filename` please BTW) assignment and the `with` statement is excessive.

File: src/tutorial/play-audio.rst [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+But we don't want it to play only a small part. of the audio. We want it to
+play all of it. How do we do that? The answer is a loop.
+
+`Context` has a method, `update()`, which update itself and the source.
```
Comment (McSinyx) [inline src/tutorial/play-audio.rst]: ReST uses double tick for fixed width text, however it's a lot better if you can use the `:py:...` like in design principles.  Moreover, please separate sentences by 2 spaces.

File: src/tutorial/play-audio.rst [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+   from palace import Device, Context, Source
+
+   with Device(), Context() as ctx:
+       with Source() as src:
+           # to be written
+
+Just like for the case of `Context`, `Source` creation
+requires a context, but here the context is passed implicitly.
+
+Decode the audio file
```
Comment (McSinyx) [inline src/tutorial/play-audio.rst]: Please use title case for headings.

File: src/tutorial/index.rst [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+   :maxdepth: 2
+
+   context
+   play-audio
+.. comment these to add later
+   Moving sources
+   Adding effects
+   Customize decoder
+   Generate sounds
+..
```
Comment (McSinyx) [inline src/tutorial/index.rst]: Please remove this last line.

File: src/tutorial/context.rst [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+       # Your code goes here
+
+Before we continue, if we only use `dev` for context creation,
+is there a shorter way to write it? The answer is yes.
+
+.. code-block::
+
+   from palace import Device, Context
+
+   with Device(), Context() as ctx:
```
Comment (McSinyx) [inline src/tutorial/context.rst]: Missing highlight, also `Context()` does not work (yet, and I don't think it's a good idea either).

**Review by McSinyx — CHANGES_REQUESTED** [review]

File: src/tutorial/context.rst [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+such as sources, buffers, and effects.
+
+Creating a Device Object
+------------------------
+
+To create a context, we must first create a device,
+since it's a parameter of the context object.
+
+To create an object, well, you just have to instantiate
+the `Device` class.
```
Comment (McSinyx) [inline src/tutorial/context.rst]: Missing reference here, actually in the entire page.

File: src/tutorial/play-audio.rst [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+   filename = 'some_audio.ogg'
+   with Device() as dev, Context(dev) as ctx:
+       with Source() as src:
+           dec = decode(filename)
+           dec.play(12000, 4, src)
+
+But we don't want it to play only a small part. of the audio. We want it to
+play all of it. How do we do that? The answer is a loop.
+
+:py:class:`Context` has a method, :py:meth:`update`, which update itself and the source.
```
Comment (McSinyx) [inline src/tutorial/play-audio.rst]: It's `` :py:meth:`Context.update` ``.

File: src/tutorial/play-audio.rst [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+   with Device() as dev, Context(dev) as ctx:
+       with Source() as src:
+           dec = decode(filename)
+           dec.play(12000, 4, src)
+           while src.playing:
+               ctx.update()
+
+If you tried this code for a song, you will find that it's a bit rush.
+That is because the source is renewed too fast. So, a simple solution
+is to :py:func:`sleep` for a while.
```
Comment (McSinyx) [inline src/tutorial/play-audio.rst]: We don't generate reference for standard modules, just use  double backtick here.

**Review by McSinyx — CHANGES_REQUESTED** [review]

File: src/tutorial/play-audio.rst (hunk 6 above)
Comment (McSinyx) [reply src/tutorial/play-audio.rst]: Ping on the sentence separation thing.

**Review by McSinyx — APPROVED** [review]

**Pull request merged**